docs/design_docs/shared_steps.md

```diff
@@ -0,0 +1,74 @@
+# Shared steps
+
+date: 2023/08/18
+
+Contributors: Carolyn Begeman, Xylar Asay-Davis
+
+## Summary
+
+The capability designed here is the ability to share steps across test cases.
+In this design document, "shared steps" refers to any step which may be used by
+multiple test cases that are available in polaris.
+
+The main motivation behind this capability is the computational expense of
+running steps that could shared across test cases multiple times. In order to
+reflect the fact that steps are shared to the user, we present a new design for
+the working directory structure. The design is successful insofar as it
+guarantees that shared steps are run once per slurm job and that the role of
+shared steps is clear to users.
+
+## Requirements
+
+### Desired: Shared steps are set up once.
+
+Ideally, steps used by multiple test cases should be run once per invocation of
+`polaris setup` or `polaris suite`. This requirement may be dropped since step
+set up is not generally resource intensive.
+
+### Requirement: Shared steps are run once.
+
+Shared steps should be run once per invocation of `polaris serial` or
+`polaris parallel`.
+
+### Requirement: Shared steps are run before steps that depend on their output.
+
+### Requirement: Object properties of shared steps are intuitive.
+
+Shared steps are not daughters of test cases.
+
+### Requirement: Working directory structure is intuitive.
+
+Shared step directories should be located at the highest level in the working directory structure where all test cases that use that step are run at or below that level.
```

Comment:
```suggestion
Shared step directories should be located at the highest level in the working directory 
structure where all test cases that use that step are run at or below that level.
```

Reply:
@cbegeman, I'm finding this requirement hard to work with in practice.  The situation where I'm having trouble is when I have a test case that builds off of another test case.  It doesn't feel intuitive to have these test cases inside of the test cases they build off of, at least not to me.

As an example, we have a `cosine_bell` test case and currently a `cosine_bell_with_viz`.  It makes more conceptual sense for these to be a the same level, and for the `cosine_bell_with_viz` to run all the steps from `cosine_bell` and then its additional viz related steps.  There are similar situations in the global ocean where it makes more sense to me for `dynamic_adjustment` and `files_for_e3sm` to be at the same level, rather than having `files_for_e3sm` inside of `dynamic_adjustment`.

@sbrus89 proposed having symlinks within a test case to the shared steps that don't reside within the test case's work directory.  This might be a more practical way of clarifying things.  A user could see all the steps in a test case clearly listed within the test case, either as subdirectories or as symlinks.

There would then be 2 requirements:
* all all shared steps in a test case either reside within the test case or have a symlink to them in the test case
* a shared step can only reside in a test case's work directory if other test cases that use those shared steps use *all* steps in that test case (i.e. subsequent test cases build off of that test case).

Reply:
@xylar I don't like having test cases that depend on one another at the same level in the directory structure. To me, the intuitive thing is having a tree with one test case branch when we set up one test case. But I defer to your judgment as the one working most closely on this. Your two requirements are an acceptable compromise to me.